src/vmm/src/arch/aarch64/regs.rs

```diff
@@ -157,265 +110,733 @@ arm64_sys_reg!(ID_AA64MMFR2_EL1, 3, 0, 0, 7, 2);
-                .get_one_reg(*id)
-                .map_err(|e| Error::GetOneReg(*id, e))?,
-        });
-    }
-
-    Ok(())
+/// Iterator over `Aarch64RegisterVec`.
+#[derive(Debug)]
+pub struct Aarch64RegisterVecIterator<'a> {
+    index: usize,
+    offset: usize,
+    vec: &'a Aarch64RegisterVec,
 }
 
-/// Returns all registers ids, including core and system
-pub fn get_all_registers_ids(vcpufd: &VcpuFd) -> Result<Vec<u64>, Error> {
-    // Call KVM_GET_REG_LIST to get all registers available to the guest. For ArmV8 there are
-    // less than 500 registers.
-    let mut reg_list = RegList::new(500).map_err(Error::Fam)?;
-    vcpufd
-        .get_reg_list(&mut reg_list)
-        .map_err(Error::GetRegList)?;
-    Ok(reg_list.as_slice().to_vec())
+impl<'a> Iterator for Aarch64RegisterVecIterator<'a> {
+    type Item = Aarch64RegisterRef<'a>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.index < self.vec.ids.len() {
+            let id = self.vec.ids[self.index];
+            let reg_size = reg_size(id) as usize;
+            let reg_ref = Aarch64RegisterRef {
+                id,
+                data: &self.vec.data[self.offset..self.offset + reg_size],
+            };
+            self.index += 1;
+            self.offset += reg_size;
+            Some(reg_ref)
+        } else {
+            None
+        }
+    }
 }
 
-/// Set the state of the system registers.
-///
-/// # Arguments
-///
-/// * `regs` - Slice of registers to be set.
-pub fn set_registers(vcpufd: &VcpuFd, regs: &[Aarch64Register]) -> Result<(), Error> {
-    for reg in regs {
-        vcpufd
-            .set_one_reg(reg.id, reg.value)
-            .map_err(|e| Error::SetOneReg(reg.id, e))?;
-    }
-    Ok(())
+/// Iterator over `Aarch64RegisterVec` with mutable values.
+#[derive(Debug)]
+pub struct Aarch64RegisterVecIteratorMut<'a> {
+    index: usize,
+    offset: usize,
+    vec: &'a mut Aarch64RegisterVec,
 }
 
-/// Get the multistate processor.
-///
-/// # Arguments
-///
-/// * `vcpu` - Structure for the VCPU that holds the VCPU's fd.
-pub fn get_mpstate(vcpufd: &VcpuFd) -> Result<kvm_mp_state, Error> {
-    vcpufd.get_mp_state().map_err(Error::GetMp)
+impl<'a> Iterator for Aarch64RegisterVecIteratorMut<'a> {
+    type Item = Aarch64RegisterRefMut<'a>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.index < self.vec.ids.len() {
+            let id = self.vec.ids[self.index];
+            let reg_size = reg_size(id) as usize;
+            // # Safety:
+            // Registers data packed in the `vec.data`
+            // The offset for each register data is a
+            // sum of sizes of previous registers sizes.
+            let data = unsafe {
+                let data = self.vec.data.as_mut_ptr();
+                let reg_data = data.add(self.offset);
+                std::slice::from_raw_parts_mut(reg_data, reg_size)
+            };
+            self.index += 1;
+            self.offset += reg_size;
```

Comment:
This should simply be `&reg_data[self.offset..self.offset + reg_size]`, or similar

Reply:
We need to return a mutable slice inside `Aarch64RegisterRefMut` and this slice is not tied to the iterator itself, but to the vec inside.

Reply:
Ahh, right the rust iterator problem. ugh. And I guess since the registers all have difference sizes we also cant use standard lib methods that usually work around this, like chunks_mut?

Reply:
Nah, `chunks_mut` won't help here.

Reply:
Actually, it seems that the mutable version of this iterator is not used anymore. Are there plans to use it in a follow up?

Reply:
it is used for the cpu templates in `vmm/src/cpu_config/aarch64/mod.rs`